File: cpp solution/string compression.cpp
class Solution {
public:
    int compress(vector<char>& chars) {
        int n = chars.size();
        int ansIndex = 0;
        int currIndex = 0;
        while (currIndex < n){
            int repeatingIndex = currIndex + 1;
            // keep moving repeating character until we get different character
            while (repeatingIndex < n && chars[repeatingIndex] == chars[currIndex]){
                repeatingIndex++;
            }
            // store the old character
            chars[ansIndex++] = chars[currIndex];
            int repeatingCharCount = repeatingIndex -  currIndex;
            // convert count into string and modify chars array with each 'count char'
            if (repeatingCharCount > 1){
                string count = to_string(repeatingCharCount);
                for (auto ch: count){
                    chars[ansIndex++] = ch;
                }
            }
            currIndex = repeatingIndex;
        }
        return ansIndex;
    }
};

File: python solution/string compression.py
# time: O(n)
class Solution:
    def compress(self, chars: List[str]) -> int:
        size = len(chars)
        ansIndex = 0
        startIndex = 0
        while startIndex < size:
            repeatingIndex = startIndex + 1
            # keep moving forward until we get different char
            while repeatingIndex < size and chars[repeatingIndex] == chars[startIndex]:
                repeatingIndex += 1
                
            # store the current char
            chars[ansIndex] = chars[startIndex]
            ansIndex += 1
            repeatingCount = repeatingIndex - startIndex
            # convert repeating count in string and modify chars array with each char
            if repeatingCount > 1:
                repeatingCount = str(repeatingCount)
                for char in repeatingCount:
                    chars[ansIndex] = char
                    ansIndex += 1
            startIndex = repeatingIndex
        return ansIndex
